mitre/codi-schema
# Update `CENSUS_DEMOG`

# Summary

* Adds a Ruby script, `generate_census_demog_sql.rb`, to generate a SQL data loading file with updated census demographic data.
  * Only imports CSV columns matching `vdw.census_demog` table columns
  * Handles erroneous data, e.g. with a scale larger than allowed by the census demog schema, by making the value to NULL.
* Corrects the precision and scale of columns within the `census_demog` table from (11,10) to (6,4)
* Removes the PRIMARYKEY contraint in the `census_demog` table which is not specified in CODI DM 4.1.5

** ❗Note:** The source CSV files nor the generated SQL insert files are included because they exceed GitHub's file size limit of 100.00 MB. Options to address this include:
- Require users to generate their own census demographic data (Current implementation)
- Provide census demog CSVs and/or SQL files out of band
- Adopt Git Large File Storage and commit census demog sql files
- Break sql files up into 11 smaller files
- Filter down to just NC data and enable option to filter by state (I like this option the best as it continues the previous precedent and I'm guessing data-owner patient populations are mostly contained to their state.)
- Other?

```shell
remote: error: File census_demog_2019_lt25.sql is 426.20 MB; this exceeds GitHub's file size limit of 100.00 MB
remote: error: File census_demog_2019_gt25.sql is 609.60 MB; this exceeds GitHub's file size limit of 100.00 MB
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
```

**TODO:**
- ~Capitalize `CHORDS_geocode` in the CSV so it matches the column name. (I could make the code case-insensitive, but I'm not sure it's the right approach or worth the effort)~
- ~Filter out any rows that are NOT null for the BLOCKGRP column~

# Testing

- Download the census_demog csv files and run `ruby generate_census_demog_sql.rb`.
- Uncomment the names of the generated sql insertion files from `build_db.rb`
- Create a new database and run `build_db.rb`
- Query the `vdw.census_demog` table to ensure the data is loaded.
  - Loading both `census_demog_2019_lt25.sql` and `census_demog_2019_gt25.sql` should insert 1,484,974 records.

# Other
## Approach Considerations

A few approaches were considered before settling on **Option A** - generating a ruby script which directly generates the SQL data loading file.

**Option A: Generate SQL Loading File with Ruby Script**
This repository makes heavy use of Ruby already and Ruby allows us to preprocess the CSV input in order to ignore columns of disinterest, set the batch size, and preprocess the CSV files into a SQL file which can be distributed.

**Option B: Load the CSV Directly with Ruby**
This method also allows processing of the input, but may be slower in practice because that processing must be done at load time. By preprocessing we can bypass the Ruby processing entirely at load time and submit the data loading query to the database directly.

**Option C: Load into Postgres with `\copy` and create a data dump**
This option was explored, but the extraneous columns in the CSV files made loading difficult. because "COPY FROM will raise an error if any line of the input file contains more or fewer columns than are expected." ([Postgres](https://www.postgresql.org/docs/current/sql-copy.html)). Pre-processing of the CSV to remove extraneous columns could alleviate this issue, but at that point we might as well generate the SQL directly.
